File: leetcode_python/82.py
# 82. 删除排序链表中的重复元素 II
# https://leetcode-cn.com/problems/remove-duplicates-from-sorted-list-ii/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
import heapq
'''
@File : 82.py
@Time : 2021/12/17 15:32:33
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Medium
'''
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if not head:
            return head
        
        dummy = ListNode(0, head)

        cur = dummy
        while cur.next and cur.next.next:
            if cur.next.val == cur.next.next.val:
                x = cur.next.val
                while cur.next and cur.next.val == x:
                    cur.next = cur.next.next
            else:
                cur = cur.next

        return dummy.next

File: leetcode_c_plus_plus/82.cpp
// 82. 删除排序链表中的重复元素 II
// https://leetcode-cn.com/problems/remove-duplicates-from-sorted-list-ii/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 82.cpp
@Time : 2022/01/01 12:44:39
@Author : YuMin Zhang
@State : Indepeedent
@Thinking :
@Tag :Medium
**/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode * dummy = new ListNode(-101, head);
        ListNode * curr = dummy;
        while(curr && curr->next && curr->next->next) {
            if(curr->next->val == curr->next->next->val){
                while(curr->next->next && curr->next->val == curr->next->next->val) {
                    curr->next->next = curr->next->next->next;
                }
                curr->next = curr->next->next;
            } else {
                curr = curr->next;
            }
        }

        ListNode * ans = dummy->next;
        delete dummy;
        return ans;

    }
};
